### PR TITLE
generalize getProtoKeystoneContext() and getKeystoneContext() functions

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
 
   ProtoKeystoneContext(int height, int time)
-      : vbkBlockHeight(height), timestampOfEndorsedBlock(time) {}
+      : blockHeight(height), timestampOfEndorsedBlock(time) {}
 
-  // A map of the Bitcoin blocks which reference (endorse a block header which
-  // is the represented block or references the represented block) to the index
-  // of the earliest
-  std::set<BlockIndex<BtcBlock>*> referencedByBtcBlocks;
+  // A map of the Endorsment blocks which reference (endorse a block header
```
Comment (blackyblack) [inline include/veriblock/blockchain/pop/fork_resolution.hpp]: Endorsement

File: include/veriblock/blockchain/pop/fork_resolution.hpp [hunk 2]
```diff
@@ -39,6 +43,129 @@ struct ComparePopScore {
...
+              if (adjustedEndorsementIndex < earliestEndorsementIndex) {
+                earliestEndorsementIndex = adjustedEndorsementIndex;
+              }
+
+              // Always break; we found a valid Bitcoin index and any
+              // future adjustedEndorsementIndex is going to be higher
+              break;
+            }  // end if
+          }    // end for
+        }      // end for
```
Comment (blackyblack) [inline include/veriblock/blockchain/pop/fork_resolution.hpp]: Could we split this code to smaller function? At least move it out of the lambda function.

File: include/veriblock/popmanager.hpp [hunk 3]
```diff
@@ -63,6 +63,18 @@ struct PopManager {
   /// commit currently uncommitted payloads
   void commit();
 
+  /**
+   * Determine the best chain of the AltBlocks in accordance with the VeriBlock
+   * forkresolution rules
+   * @param AltBlock chain1, AltBlock chain2
+   * @return '-1' if chain1 is better, '1' if chain is better, '0' if they are
```
Comment (blackyblack) [inline include/veriblock/popmanager.hpp]: If chain2 is better

File: include/veriblock/popmanager.hpp [hunk 4]
```diff
@@ -63,6 +63,18 @@ struct PopManager {
...
+   * the same
+   * @note chain1 and chain2 are being consindered as a forks not a full chains
```
Comment (blackyblack) [inline include/veriblock/popmanager.hpp]: As forks not full chains

File: include/veriblock/popmanager.hpp [hunk 5]
```diff
@@ -63,6 +63,18 @@ struct PopManager {
...
+   * form the genesis block, they should start at the common block
```
Comment (blackyblack) [inline include/veriblock/popmanager.hpp]: From the genesis

File: src/popmanager.cpp [hunk 6]
```diff
@@ -118,4 +118,9 @@ bool PopManager::hasUncommittedChanges() const noexcept {
   return !uncommitted_.empty();
 }
 
+int PopManager::compareTwoBranches(const Chain<AltBlock>&,
+                                   const Chain<AltBlock>&) {
+  return 0;
+}
```
Comment (blackyblack) [inline src/popmanager.cpp]: Why this dummy function?

Comment (Mr-Leshiy) [reply src/popmanager.cpp]: I will implement fork resolution algorithm for the altcoin later, now I just add a signature of this function.

File: include/veriblock/popmanager.hpp [hunk 7]
```diff
@@ -63,6 +63,18 @@ struct PopManager {
   /// commit currently uncommitted payloads
   void commit();
 
+  /**
+   * Determine the best chain of the AltBlocks in accordance with the VeriBlock
+   * forkresolution rules
+   * @param AltBlock chain1, AltBlock chain2
+   * @return '-1' if chain1 is better, '1' if chain2 is better, '0' if they are
+   * the same
+   * @note chain1 and chain2 are being consindered as forks not a full chains
```
Comment (blackyblack) [inline include/veriblock/popmanager.hpp]: Not full chains

**Pull request merged**